File: src/models/anthropic.js
import Anthropic from '@anthropic-ai/sdk';
import { BaseModel } from './base.js';
import { config } from '../config.js';

export class AnthropicModel extends BaseModel {
  constructor() {
    super('Anthropic');
    this.client = new Anthropic({
      apiKey: config.anthropic.apiKey
    });
  }

  async sendMessage(messages, onChunk) {
    try {
      const startTimeMs = Date.now();
      const stream = await this.client.messages.create({
        model: 'claude-sonnet-4-20250514',
        max_tokens: 4096,
        messages: this.formatMessages(messages),
        stream: true
      });
      
      let fullContent = '';
      let usage = null;
      
      for await (const chunk of stream) {
        if (chunk.type === 'content_block_delta') {
          fullContent += chunk.delta.text;
          if (onChunk) {
            onChunk(this.name, chunk.delta.text);
          }
        } else if (chunk.type === 'message_delta' && chunk.usage) {
          usage = chunk.usage;
        }
      }
      const responseTimeSeconds = (Date.now() - startTimeMs) / 1000;

      // Normalize usage keys and compute totals if available
      let normalizedUsage = null;
      if (usage) {
        const input = usage.input_tokens ?? usage.prompt_tokens ?? 0;
        const output = usage.output_tokens ?? usage.completion_tokens ?? 0;
        const total = (usage.total_tokens != null) ? usage.total_tokens : (input + output || undefined);
        normalizedUsage = {
          input_tokens: usage.input_tokens ?? undefined,
          output_tokens: usage.output_tokens ?? undefined,
          total_tokens: total,
          // also provide prompt/completion for generic handling
          prompt_tokens: usage.prompt_tokens ?? usage.input_tokens ?? undefined,
          completion_tokens: usage.completion_tokens ?? usage.output_tokens ?? undefined
        };
      }

      return {
        model: this.name,
        content: fullContent,
        usage: normalizedUsage || usage,
        response_time: responseTimeSeconds
      };
    } catch (error) {
      // Enhanced error handling with more context
      console.error('🚨 ANTHROPIC ERROR DEBUG:', error);

      let detailedError;
      const errorMsg = error.message?.toLowerCase() || '';
      const statusCode = error.status || error.response?.status;
      const errorType = error.error?.type || '';

      if (statusCode === 401) {
        detailedError = `Authentication failed (401): Check API key validity`;
      } else if (statusCode === 429) {
        detailedError = `Rate limited (429): Too many requests, try again later`;
      } else if (statusCode === 500 || statusCode === 502 || statusCode === 503) {
        detailedError = `Anthropic server error (${statusCode}): Service temporarily unavailable`;
      } else if (statusCode === 400) {
        detailedError = `Invalid request (400): ${error.message}`;
      } else if (errorType === 'overloaded_error') {
        detailedError = `Service overloaded: Anthropic servers are experiencing high demand`;
      } else if (errorMsg.includes('connection') || errorMsg.includes('timeout') || errorMsg.includes('network') || errorMsg.includes('enotfound') || errorMsg.includes('econnreset')) {
        detailedError = `Network connection error: Unable to reach Anthropic servers - ${error.message}`;
      } else if (errorMsg.includes('api') && errorMsg.includes('key')) {
        detailedError = `API key error: ${error.message}`;
      } else if (errorMsg.includes('quota') || errorMsg.includes('billing')) {
        detailedError = `Quota/billing issue: ${error.message}`;
      } else {
        detailedError = `Anthropic API error${statusCode ? ` (${statusCode})` : ''}: ${error.message}`;
      }

      return {
        model: this.name,
        error: detailedError
      };
    }
  }

  formatMessages(messages) {
    return messages.map(msg => ({
      role: msg.role === 'system' ? 'assistant' : msg.role,
      content: msg.content
    }));
  }
}

File: src_python/models/anthropic.py
import anthropic
from typing import List, Dict, Any, Optional, Callable
from .base import BaseModel
from ..config import config


class AnthropicModel(BaseModel):
    def __init__(self):
        super().__init__('Anthropic')
        self.client = anthropic.AsyncAnthropic(
            api_key=config['anthropic']['api_key']
        )
    
    async def send_message(
        self, 
        messages: List[Dict[str, str]], 
        on_chunk: Optional[Callable[[str, str], None]] = None
    ) -> Dict[str, Any]:
        try:
            formatted_messages = self.format_messages(messages)
            
            if on_chunk:
                full_content = ''
                usage = None
                
                async with self.client.messages.stream(
                    model='claude-3-5-sonnet-20241022',
                    max_tokens=4096,
                    messages=formatted_messages
                ) as stream:
                    async for chunk in stream:
                        if hasattr(chunk, 'type'):
                            if chunk.type == 'content_block_delta':
                                full_content += chunk.delta.text
                                on_chunk(self.name, chunk.delta.text)
                            elif chunk.type == 'message_delta' and hasattr(chunk, 'usage'):
                                usage = chunk.usage
                
                return {
                    'model': self.name,
                    'content': full_content,
                    'usage': usage.__dict__ if usage else None
                }
            else:
                response = await self.client.messages.create(
                    model='claude-3-5-sonnet-20241022',
                    max_tokens=4096,
                    messages=formatted_messages
                )
                
                return {
                    'model': self.name,
                    'content': response.content[0].text,
                    'usage': {
                        'input_tokens': response.usage.input_tokens,
                        'output_tokens': response.usage.output_tokens,
                        'total_tokens': response.usage.input_tokens + response.usage.output_tokens
                    } if hasattr(response, 'usage') else None
                }
        except Exception as e:
            return {
                'model': self.name,
                'error': str(e)
            }
    
    def format_messages(self, messages: List[Dict[str, str]]) -> List[Dict[str, str]]:
        return [
            {
                'role': 'assistant' if msg['role'] == 'system' else msg['role'],
                'content': msg['content']
            }
            for msg in messages
        ]